src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +143,161 @@ internal class DiagnosticComputer
+                //    and the high priority computeTask in Step 4 below.
+                CancelNormalPriorityTasks();
+
+                // Step 4:
+                //  - Execute the core 'computeTask' for diagnostic computation.
+                return await computeTask.ConfigureAwait(false);
+            }
+            finally
+            {
+                // Step 5:
+                //  - Remove the 'computeTask' from the list of current executing high priority tasks.
+                lock (s_gate)
+                {
+                    Debug.Assert(s_highPriorityComputeTasks.Contains(computeTask));
+                    s_highPriorityComputeTasks = s_highPriorityComputeTasks.Remove(computeTask);
+                }
+            }
+
+            static void CancelNormalPriorityTasks()
+            {
+                // Fire cancellation on the current token source and replace it with a new token source for future normal priority requests.
+                ImmutableHashSet<CancellationTokenSource> cancellationTokenSources;
+                lock (s_gate)
+                {
+                    cancellationTokenSources = s_normalPriorityCancellationTokenSources;
+                }
+
+                foreach (var  cancellationTokenSource in cancellationTokenSources)
+                {
+                    try
+                    {
+                        cancellationTokenSource.Cancel();
+                    }
+                    catch (ObjectDisposedException)
+                    {
+                    }
+                }
+            }
+        }
+
+        private async Task<SerializableDiagnosticAnalysisResults> GetNormalPriorityDiagnosticsAsync(
+            IEnumerable<string> analyzerIds,
+            bool reportSuppressedDiagnostics,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Step 1:
+                //  - Normal priority task must wait for all the executing high priority tasks to complete
+                //    before beginning execution.
+                await WaitForHighPriorityTasksAsync(cancellationToken).ConfigureAwait(false);
+
+                using var cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                lock (s_gate)
+                {
+                    s_normalPriorityCancellationTokenSources = s_normalPriorityCancellationTokenSources.Add(cancellationTokenSource);
+                }
+
+                // Step 2:
+                //  - Create the core 'computeTask' for computing diagnostics.
+                var computeTask = GetDiagnosticsAsync(
+                    analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo,
+                    cancellationTokenSource.Token)
```

Comment:
no point making the task here just to await it in the try/catch  just inline this